### PR TITLE
8293017: Improve hash calculation parallelism in jmod/jlink

### PR DESCRIPTION
**Pull request by shipilev**
`jmod`/`jlink` are executed during build time to produce the `jmod` and the base modules image. On slow hardware (Raspberry Pi -class, for example) and/or slow VMs (debug, interpreter-only, for example) this takes a while. Profiling shows the considerable time is spent on hashing modules either for writing out the ModuleHash attribute or for verifying the hashes are good.

Those paths can be parallelized to make them quite faster.

The major contributors to module hashing are `java.base`, `jdk.desktop` and `jdk.localedata`, so we have a significant opportunity for parallelism here. 

Motivational improvements on `make clean-images images`:

```
# x86_64 Server, release

# Baseline
real	0m11.895s
user	1m4.526s
sys	0m10.715s

# Patched
real	0m10.701s   ; <--- 1.1x improvement
user	1m5.097s
sys	0m11.260s

# x86_64 Zero, release

# Baseline
real	5m20.335s
user	7m7.791s
sys	0m7.258s

# Patched
real	2m23.551s  ; <--- 2.23x improvement
user	7m14.442s
sys	0m7.856s
```

Additional testing: 
  - [x] Linux x86_64 fastdebug, `java/util/jar`
  - [x] Linux x86_64 fastdebug, `tools/jmod`
  - [x] Linux x86_64 fastdebug, `tools/jlink`
  - [x] Linux x86_64 fastdebug `tier1`

---------
### Progress
- [ ] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8293017](https://bugs.openjdk.org/browse/JDK-8293017): Improve hash calculation parallelism in jmod/jlink

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/10060/head:pull/10060` \
`$ git checkout pull/10060`

Update a local copy of the PR: \
`$ git checkout pull/10060` \
`$ git pull https://git.openjdk.org/jdk pull/10060/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 10060`

View PR using the GUI difftool: \
`$ git pr show -t 10060`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/10060.diff">https://git.openjdk.org/jdk/pull/10060.diff</a>

</details>

### CODE REVIEW

**Comment by AlanBateman** [comment]
This issue will require discussion as it potentially impacts usage at run-time when the resolver runs at startup or when creating module layers. I also wonder if it has an impact on JDK builds where many jobs are running jmod at the same time. It may be that this has to be configured by a system property so that it can be enabled only in the JDK build.

Also just to say that Claes removed all the lambda expressions from this code in JDK 9 to help with startup as that will have to checking too. Also if we are changing this code then we should try to keep the style/conventions/line-lengths consistent where possible.

File: src/java.base/share/classes/jdk/internal/module/ModuleHashes.java [hunk 1]
```diff
@@ -162,16 +166,28 @@ static byte[] computeHash(Supplier<ModuleReader> supplier, String algorithm) {
...
     static ModuleHashes generate(Set<ModuleReference> mrefs, String algorithm) {
-        Map<String, byte[]> nameToHash = new TreeMap<>();
-        for (ModuleReference mref : mrefs) {
-            try (ModuleReader reader = mref.open()) {
-                byte[] hash = computeHash(reader, algorithm);
-                nameToHash.put(mref.descriptor().name(), hash);
-            } catch (IOException ioe) {
-                throw new UncheckedIOException(ioe);
+        Map<String, byte[]> nameToHash = new ConcurrentHashMap<>();
+        mrefs.stream().parallel().forEach(mref -> {
```
Comment (stsypanov) [inline src/java.base/share/classes/jdk/internal/module/ModuleHashes.java:170]: [AFAIK](https://stackoverflow.com/questions/28985704/parallel-stream-from-a-hashset-doesnt-run-in-parallel) streams taken from HashSet might have weak parallelism, so maybe it's worth wrapping `mrefs` into an ArrayList?

**Comment by shipilev** [comment]
> This issue will require discussion as it potentially impacts usage at run-time when the resolver runs at startup or when creating module layers. 

Startup with/without CDS:

```
$ perf stat -r 100 build/linux-x86_64-server-release/images/jdk/bin/java Hello

# Baseline
 With CDS: 0.0323543 +- 0.0000789 seconds time elapsed  ( +-  0.24% )
 Without CDS: 0.071471 +- 0.000500 seconds time elapsed  ( +-  0.70% )

# Patched
 With CDS: 0.0328651 +- 0.0000780 seconds time elapsed  ( +-  0.24% )
 Without CDS:  0.072745 +- 0.000472 seconds time elapsed  ( +-  0.65% )
```

So there seem to be some minor impact, but nothing drastic. I would be very surprised if we actually verified the module hashes at startup! That would take minutes to startup on Zero, as observed by its build-time checking. And we don't see that, apparently.

> I also wonder if it has an impact on JDK builds where many jobs are running jmod at the same time. It may be that this has to be configured by a system property so that it can be enabled only in the JDK build.

The normal build already runs a lot of `jmod`-s at once. But that would not help when you have the fat `jmod java.base` at the end of the build, where no external parallelism exists. At that point, internal `Resolver` parallelism helps a lot.

> Also just to say that Claes removed all the lambda expressions from this code in JDK 9 to help with startup as that will have to checking too. Also if we are changing this code then we should try to keep the style/conventions/line-lengths consistent where possible.

I reverted some brace changes.

**Comment by AlanBateman** [comment]
> I would be very surprised if we actually verified the module hashes at startup! That would take minutes to startup on Zero, as observed by its build-time checking. And we don't see that, apparently.

Module hashes are checked at startup when deploying with tightly coupled modules on the module path. You should see some examples if you run the tests test/jdk/tools/jmod/hashes. It's the same thing when configuring the Configuration for module layers at run-time. In the JDK build, the hashes are added at packaging time so you don't see checking when using exploded builds. At run-time, without an application module, then modules are loaded from the run-time image where we can safety skip the checking.

I have no objection to using a parallel-stream be opt-in by the JDK build but I don't think it can be default because it impacts the behavior of standard APIs and out-of-the-box behavior.

**Pull request closed without merging**